src/Symfony/Component/Messenger/Worker.php

```diff
@@ -48,7 +65,81 @@ public function run()
                 return;
             }
 
-            $this->bus->dispatch($envelope->with(new ReceivedStamp()));
+            $this->dispatchEvent(
+                WorkerMessageHandlingEvent::class,
+                new WorkerMessageHandlingEvent($envelope, $this->receiverName)
+            );
+
+            $message = $envelope->getMessage();
+            $context = [
+                'message' => $message,
+                'class' => \get_class($message),
+            ];
+
+            try {
+                $this->bus->dispatch($envelope->with(new ReceivedStamp()));
+            } catch (\Throwable $throwable) {
+                $shouldRequeue = $this->shouldRequeue($throwable, $envelope);
+
+                $this->dispatchEvent(
+                    WorkerMessageFailedEvent::class,
+                    new WorkerMessageFailedEvent($envelope, $this->receiverName, $throwable, $shouldRequeue)
+                );
```

Comment:
The events are currently dispatched before ack/reject/retry on the queue. That's subjective, and either order could, in theory, cause a situation where one fails and so the other doesn't run (e.g. some listener throws an exception, so the retry never happens, or, if we reverse, the retry fails due to a network connection, then the event is never dispatched). Not sure if we need to be thinking about this level of failure. Catching exceptions makes things harder to debug/know about when they go wrong.